### ISSUE TITLE
Optimize deletion of read docs for databases with many users

### ISSUE THREAD

**Issue by dianabarsan** [medic/medic#5420]
A sister of https://github.com/medic/medic/issues/5395 .

When Sentinel receives a delete, it attempts to delete the `read` docs from every user database. It doesn't batch and executes all requests in parallel. 

If there are enough users, it sometimes crashes CouchDB (I tested with 1000 users, on 3.3.x with Couchdb 2.3), other times just failing the operation. 

```
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]    { 'cache-control': 'must-revalidate',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]      uri: 'http://XXXXXX:XXXXXX@localhost:5984/medic-user-tom266-meta/_all_docs',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]    { method: 'POST',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]   reason: 'function_clause',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]   [message]: 'function_clause',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]     at Request.emit (events.js:211:7)
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]   description: 'couch returned 500' } 
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]   errid: 'non_200',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]      'x-couch-request-id': 'd98a21ebc8',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]      connection: 'close',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]         accept: 'application/json' },
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]      headers: 
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]   request: 
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]     at Request.self.callback (/srv/software/medic-sentinel/md5-6b4BPRyG927vstXfB5ChAg==/node_modules/nano/node_modules/request/request.js:185:22)
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]   headers: 
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]       { 'content-type': 'application/json',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]   error: 'unknown_error',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]   [stack]: 'Error: function_clause\n    at Request._callback (/srv/software/medic-sentinel/md5-6b4BPRyG927vstXfB5ChAg==/node_modules/nano/lib/nano.js:252:15)\n    at Request.self.callback (/srv/software/medic-sentinel/md5-6b4BPRyG927vstXfB5ChAg==/node_modules/nano/node_modules/request/request.js:185:22)\n    at emitTwo (events.js:126:13)\n    at Request.emit (events.js:214:7)\n    at Request.<anonymous> (/srv/software/medic-sentinel/md5-6b4BPRyG927vstXfB5ChAg==/node_modules/nano/node_modules/request/request.js:1157:10)\n    at emitOne (events.js:116:13)\n    at Request.emit (events.js:211:7)\n    at IncomingMessage.<anonymous> (/srv/software/medic-sentinel/md5-6b4BPRyG927vstXfB5ChAg==/node_modules/nano/node_modules/request/request.js:1079:12)\n    at Object.onceWrapper (events.js:313:30)',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]     at Object.onceWrapper (events.js:313:30)
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34] 2019-02-18 18:21:34 ERROR: Error cleaning up deleted doc: { Error: function_clause
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]      statusCode: 500,
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]      'x-couchdb-body-time': '0',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]      'x-couch-stack-hash': '1041696419',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]      date: 'Mon, 18 Feb 2019 18:21:34 GMT',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]      qs: { include_docs: true },
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]   statusCode: 500,
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]   scope: 'couch',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]   name: 'Error',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]     at IncomingMessage.<anonymous> (/srv/software/medic-sentinel/md5-6b4BPRyG927vstXfB5ChAg==/node_modules/nano/node_modules/request/request.js:1079:12)
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]     at emitOne (events.js:116:13)
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]     at Request.<anonymous> (/srv/software/medic-sentinel/md5-6b4BPRyG927vstXfB5ChAg==/node_modules/nano/node_modules/request/request.js:1157:10)
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]     at Request._callback (/srv/software/medic-sentinel/md5-6b4BPRyG927vstXfB5ChAg==/node_modules/nano/lib/nano.js:252:15)
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]      uri: 'http://XXXXXX:XXXXXX@localhost:5984/medic-user-tom266-meta/_all_docs' },
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]      'content-type': 'application/json',
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]      body: '{"keys":["read:report:messages-ge","read:message:messages-ge"]}' },
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]   ref: 1041696419,
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]     at Request.emit (events.js:214:7)
Feb 18 20:21:34 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-18 18:21:34]     at emitTwo (events.js:126:13)
```

**Comment by dianabarsan** [comment]
@garethbowen please advise where we should schedule this.

**Comment by garethbowen** [comment]
I'm not sure how often deletes happen but crashing CouchDB is not ok... I'll look at getting this fixed today.

**Comment by garethbowen** [comment]
Merged into `master` and backported to `3.4.x` and `3.3.x` and ready for AT.

For AT: Using an instance with many user dbs log in as an offline user and mark a report as read and sync that to your remote user db. Log in as an admin and delete the doc. Watch the sentinel log for the change to be processed. It may take a while but it should check every user db and delete all the read docs. Ensure sentinel and couchdb don't crash. Ensure the read doc is deleted from the offline users db so it's no longer counted as a read doc.

**Comment by newtewt** [comment]
@garethbowen  I haven't seen anything yet to indicate it's going over after deleting. By marking the doc are read you mean Clicking Review then clicking Correct?

**Comment by garethbowen** [comment]
You mark a doc as read by selecting it in the LHS. Clicking "review" is a different workflow. If a doc is unread it has this little blue bar at the top right:

![screenshot from 2019-02-19 16-13-20](https://user-images.githubusercontent.com/2064938/52987753-555ad900-3461-11e9-8329-c3b0ca75a898.png)

**Comment by ngaruko** [comment]
LGTM
Sentinel logs show every user db,
![image](https://user-images.githubusercontent.com/6979995/52997351-32dab700-3485-11e9-9c77-5c227ccc3a10.png)

and a final tombstone saving ...
```Feb** 19 20:09:33 gamma-dev gamma-dev-medic-sentinel-logs:  (gamma-dev-546bfc466c-dkgmd) | [2019-02-19 07:09:33] 2019-02-19 07:09:33 INFO: saving tombstone for 11096e274eadfa89fbffb16d940058b0```

No crashes
